### ISSUE TITLE
clean up FileTransfer::checkPath()

### ISSUE THREAD

**Issue by bradbulger** [backdrop/backdrop-issues#5190]
The code in FileTransfer::checkPath() appears to be doing more than it really does. If the `$path` argument does not start with the literal absolute filesystem path that matches the jail property - which in practice is cwd() from index.php - then the function will only pass values in unlikely or accidental ways.

It would be good to settle open questions - what is it meant to pass, what is it meant to fail, does it actually support stream wrappers? - and fix the code to match the answers.

This is a spinoff of #3008.

**Comment by indigoxela** [comment]
It's maybe helpful to point out, what was discovered in https://github.com/backdrop/backdrop-issues/issues/3008:

If the path doesn't start with the actual real path, this can lead to unpredictable results, as it depends on the length of the real and the symlinked base path:

```
$path = '/var/www/bdroptesting/html/modules/devel';// The unresolved path, "html" is a symlink.
$full_jail = '/var/www/bdroptesting/html_real';// The length of the resolved base path differs in this case and could be something totally different anyway.
...
...substr($this->chroot . $path, 0, strlen($full_jail)) // '/var/www/bdroptesting/html/modu' -> Obviously wrong, can never get resolved by backdrop_realpath.
```
[Link to function](https://github.com/backdrop/backdrop/blob/1.x/core/includes/filetransfer/filetransfer.inc#L169)

**Comment by indigoxela** [comment]
I'm not sure about the issue label. IMO it's either a bug or a task, but rather not a feature request. But I'll leave it as-is until we clarified the direction here a bit more.

**Comment by klonos** [comment]
> I'm not sure about the issue label. IMO it's either a bug or a task...

Either of those can get in with a bug fix release. The way the issue title is phrased, it is a task. But if you changed the issue title to something like this:

> If the path in FileTransfer::checkPath() doesn't start with the actual real path, it can lead to unpredictable results

...then it's a bug 😉

**Comment by bradbulger** [comment]
When does authorize.php run? There are circumstances where it gets the elements of authorize_filetransfer_form, which seems like it might be creating an instance of a FileTransfer class with the jail property set to its value of BACKDROP_ROOT.

Is this an appropriate place to be asking these kinds of questions?

**Comment by indigoxela** [comment]
> Is this an appropriate place to be asking these kinds of questions?

Anything that clarifies the whole process is an appropriate question. :wink:

Here's the part of code in authorize.php including the comment, that (again) brings us back to issue #1297:
```
/**
 * Defines the root directory of the Backdrop installation.
 *
 * The dirname() function is used to get path to Backdrop root folder, which
 * avoids resolving of symlinks. This allows the code repository to be a symlink
 * and hosted outside of the web root. See issue #1297.
 */
define('BACKDROP_ROOT', dirname(dirname($_SERVER['SCRIPT_FILENAME'])));

// Change the directory to the Backdrop root.
chdir(BACKDROP_ROOT);
```

**Comment by indigoxela** [comment]
Hm, I wonder, why it's necessary to define BACKDROP_ROOT with the symlink included. Maybe it works to just use the value of `dirname(dirname($_SERVER['SCRIPT_FILENAME']))` for chdir(), but still resolve symlinks for the constant? Need to try that out.

**Comment by bradbulger** [comment]
I don't think it is necessary. I think it just worked at the time for the specific issue they were considering. I see no reason why BACKDROP_ROOT ever needs to be defined as a symlink. The chdir() would put you into the same place either way.

All of that said, the code in checkPath() is still weird, and I still am trying to figure out any circumstance where a FileTransfer object might be created with a jail being anything other than the realpath to the location of index.php.

**Comment by indigoxela** [comment]
A little off topic, as we try to focus on weirdness in checkPath()... To better understand the necessity of the breaking change in #1297, I set up following:

Docroot is /var/www/bdroptesting/html
Core is a symlink to ../core_outside/

In the core directory I added a php file with following content and called it in a browser:
```
<?php
var_dump(getcwd()); // /var/www/bdroptesting/core_outside

$foo = dirname(dirname($_SERVER['SCRIPT_FILENAME']));
var_dump($foo); // /var/www/bdroptesting/html
chdir($foo);

var_dump(getcwd()); // /var/www/bdroptesting/html
?>
```
getcwd() would cause a problem, as files and settings.php are supposed to be in /var/www/bdroptesting/html and not next to "core_outside". I guess, we have to live with that oddness in authrorize.php, install.php and the other files. :shrug: 

Back to topic:

I've no idea, how to figure out, what checkPath() is actually supposed to do and what is odd stuff that could get improved. I see potential for improvement, though.

I took a look at the D8 code and found similar stuff in core/lib/Drupal/Core/FileTransfer/FileTransfer.php
Their class FileSystem also deals with stream wrappers. No real answer to our question, I know.

**Comment by bradbulger** [comment]
i can create a separate issue about the definitions of BACKDROP_ROOT if that would be preferable. i think standardizing the value would be a good idea.
